program/src/processor.rs

```diff
@@ -6365,58 +6304,902 @@ impl Processor {
+                    .checked_mul(delist_price)
+                    .unwrap()
+                    .checked_mul(ONE_I80F48 - delist_info.liquidation_fee)
+                    .unwrap();
+
+                // Send that amount to liqor's token account
+                checked_change_net(
+                    delist_root_bank_cache,
+                    &mut delist_node_bank,
+                    &mut liqee_ma,
+                    liqee_mango_account_ai.key,
+                    delist_index,
+                    -delist_transfer,
+                )?;
+                delist_net = liqee_ma.get_net(delist_root_bank_cache, delist_index);
+
+                invoke_transfer(
+                    token_prog_ai,
+                    delist_vault_ai,
+                    liqor_delist_token_account_ai,
+                    signer_ai,
+                    &[&signers_seeds],
+                    delist_transfer.to_num::<u64>(),
+                )?;
+
+                // Transfer quote token from liqor to liqee
+                transfer_token_internal(
+                    quote_root_bank_cache,
+                    &mut quote_node_bank,
+                    &mut liqor_ma,
+                    &mut liqee_ma,
+                    liqor_mango_account_ai.key,
+                    liqee_mango_account_ai.key,
+                    QUOTE_INDEX,
+                    quote_transfer,
+                )?;
+
+                mango_emit_heap!(WithdrawLog {
+                    mango_group: *mango_group_ai.key,
+                    mango_account: *liqor_mango_account_ai.key,
+                    owner: liqor_ma.owner,
+                    token_index: delist_index as u64,
+                    quantity: delist_transfer.to_num::<u64>(),
+                });
+
+                mango_emit_heap!(LiquidateTokenAndTokenLog {
+                    mango_group: *mango_group_ai.key,
+                    liqee: *liqee_mango_account_ai.key,
+                    liqor: *liqor_mango_account_ai.key,
+                    asset_index: delist_index as u64,
+                    liab_index: QUOTE_INDEX as u64,
+                    asset_transfer: delist_transfer.to_bits(),
+                    liab_transfer: quote_transfer.to_bits(),
+                    asset_price: delist_price.to_bits(),
+                    liab_price: ONE_I80F48.to_bits(),
+                    bankruptcy: liqee_ma.is_bankrupt
+                });
+            }
+        } else if delist_net <= NEG_ONE_I80F48 {
+            // deposits are zero, borrows are positive
+            // Transfer delisting token from liqor's token account to offset liqee's borrows
+            // Transfer quote token from liqee to liqor plus fee
+
+            let max_liquidate_amount: u64 =
+                TokenAccount::load(liqor_delist_token_account_ai)?.amount.min(max_liquidate_amount);
+
+            let delist_transfer = delist_net
+                .checked_abs()
+                .unwrap()
+                .checked_ceil()
```

Comment:
why ceil() here? At first I thought you want `delist_net` to end up between 0 and 1. But it looks like -1 to 1 is ok for dust. Then wouldn't flooring here make more sense?